### PR TITLE
apiextensions: Benchmark escaping in SchemaHas and pool Schemas

### PR DESCRIPTION
**Pull request by apelisse**
SchemaHas needs to escape the apiextensions.JSONSchemaProps pointer
since it's calling an arbitrary predicate function that can keep a copy
of the pointer (even though it shouldn't, but the compiler can't know
that). The leak is resulting in a fair amount of memory allocation when
installing many CRDs in a row (about 3% of the memory allocated happens
in this method).

Using a `sync.Pool` to re-use the buffers that have to escape in the
predicate significantly reduces the number of allocations needed to run
the SchemaHas method, as shown in the following example:
    
```
> benchstat old.bench new.bench
name         old time/op    new time/op    delta
SchemaHas-8    11.0ms ± 0%     2.1ms ± 1%   -80.60%  (p=0.008 n=5+5)
    
name         old alloc/op   new alloc/op   delta
SchemaHas-8    37.5MB ± 0%     0.0MB ± 0%  -100.00%  (p=0.008 n=5+5)
    
name         old allocs/op  new allocs/op  delta
SchemaHas-8     73.1k ± 0%      0.0k       -100.00%  (p=0.008 n=5+5)
```

That's not the most critical thing we have in the apiserver installing CRDs, but that's the best ROI since it's very easy to fix :-)

#### What type of PR is this?

/kind cleanup

#### Which issue(s) this PR fixes:
Fixes stuff

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by apelisse** [comment]
/triage accepted
/sig wg-api-expresison
/cc @lavalamp @alexzielenski @Jefftree 
/priority important-soon
I don't see how that's an API change?
/remove-kind api-change

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go [hunk 1]
```diff
@@ -1364,6 +1365,22 @@ func HasSchemaWith(spec *apiextensions.CustomResourceDefinitionSpec, pred func(s
 	return false
 }
 
+var schemaPool = sync.Pool{
+	New: func() any {
+		return new(apiextensions.JSONSchemaProps)
+	},
+}
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:1372]: Keeping this global means we keep memory around from the high watermark forever. Would making the pool local to each top-level `SchemaHas` call (would require a tiny bit of refactoring) significantly decrease the huge gains you posted?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go [hunk 2]
```diff
@@ -1364,6 +1365,22 @@ func HasSchemaWith(spec *apiextensions.CustomResourceDefinitionSpec, pred func(s
 	return false
 }
 
+var schemaPool = sync.Pool{
+	New: func() any {
+		return new(apiextensions.JSONSchemaProps)
```
Comment (lavalamp) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:1370]: Will all of these objects be about the same size?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go (hunk 1 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:1372]: No, pools release unused objects after (IIRC) 2 GC passes.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go (hunk 2 above)
Comment (apelisse) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:1370]: Their size is known at compile time?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go (hunk 1 above)
Comment (alexzielenski) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:1372]: > The store scales under load (when many goroutines are actively printing) and shrinks when quiescent.

Was hard to find a reference in docs to this and this was closest I could get. After a bit a research it does seem thats the implementation. Thanks for the correction!

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go (hunk 2 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:1370]: I looked at the type and now I'm confused? There's a ton of pointers etc in it, are we clearing everything before storing it (or before reuse)?

Comment (lavalamp) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:1370]: OK I actually looked down below and I'm surprised that actually avoids an allocation but if so it's pretty cool!

**Comment by lavalamp** [comment]
OK I understand now

/approve
/lgtm

**Pull request merged**